### PR TITLE
Serve po.js in the Python Bridge

### PR DESCRIPTION
**Pull request by jelly**
Implement serving po.js files based on the Accept-Language header. So far this only implements it for normal pages, for the shell we need to rewrite the currently used `/*/po.js` implementation into something else (follow up).

Screenshot from 2023-01-11 18-38-59

### CODE REVIEW

**Comment by jelly** [comment]
The shell is not translated as the translations are loaded through globbing "/*/po.js" which is something we might not want to support at all. Leaving this for a follow up: https://issues.redhat.com/browse/COCKPIT-852

**Comment by allisonkarlitskaya** [comment]
Answers to your questions:
 - the yielding of the variants also covers gzipped vs. non-gzipped, but I agree that trying locale variants for anything not called `po.js` is silly.  On the other hand, it's not slow either: we keep a hash table of the files that are present, so it's not like we're doing 1000 syscalls.
 - as for unit testing, I see no problem with reaching directly into the code to directly call the method, even without an instance.
 - about globbing: I'd like to avoid that, ya.  This is something we need to figure out a better approach for, and is sort of the motivating factor in my desire to rethink how we approach translation.  Long story short: we should probably serve this information as part of manifests.json.  That might look like including the translations for the relevant strings directly into manifest.json of each package at build-time, but I'm not sure yet if we want to do that.  In short: we need to talk about this, and I agree that it's OK to leave it unaddressed for now.

**Comment by jelly** [comment]
Sadly, this PR will not make any new tests pass as we check the language with `cockpit.language` which interestingly is set by `cockpit.locale` which is interestingly called in `po.js`. So as we don't have translations for the shell yet.

PS: How the hell does this globbing magic even work of loading "/po/*/po.js", does it really concatenate all po files together for all "applications"? I guess I rather not found out :ghost:

File: src/cockpit/packages.py [hunk 1]
```diff
@@ -43,6 +44,29 @@ def directory_items(path):
...
+    Returns an ordered list of (language, weight) pairs.
+
+    https://tools.ietf.org/html/rfc7231#section-5.3.5
+    """
+
+    accept_languages = accept_language.split(',')
+    locales = []
+    for index, language in enumerate(accept_languages):
+        if ';' in language:
+            locale, weight = language.split(';')
```
Comment (allisonkarlitskaya) [inline src/cockpit/packages.py:59]: According to the spec, the weight defaults to 1 if it's not present.  Look under "[5.3.1](https://www.rfc-editor.org/rfc/rfc7231#section-5.3.1).  Quality Values"

Something like this seems sensible:

```python
locale, _, weight = accept_language.partition('?q=')
weight = float(weight or 1)
```

File: src/cockpit/packages.py [hunk 2]
```diff
@@ -43,6 +44,29 @@ def directory_items(path):
...
+            locale = locale.strip()
+            weight = weight.split('=')[-1].strip()
+            locales.append((locale, weight))
+        else:
+            # Retain given locales order by assiging a fake weight
+            locales.append((language.strip(), str(1 - (index / 100))))
+
+    return list(reversed(sorted(locales, key=lambda k: k[1])))
```
Comment (allisonkarlitskaya) [inline src/cockpit/packages.py:67]: `sorted` takes a `reverse=` kwarg.  Your use of `reversed()` here reverses the original order, but `reverse=` would preserve it, removing the need to do the indexing trick above.

I also don't see the point of returning the tuples ­— you really just care about the ordered list.

File: src/cockpit/packages.py [hunk 3]
```diff
@@ -159,12 +183,18 @@ def filename_variants(filename, locale):
...
     def negotiate_file(self, path, headers):
         dirname, sep, filename = path.rpartition('/')
-        locale = headers.get('Accept-Language', '').split(',')[0].strip()  # obviously fix this
-
-        for variant in self.filename_variants(filename, locale):
-            logger.debug('consider variant %s', variant)
-            if Path(f'{dirname}{sep}{variant}') in self.files:
-                return f'{dirname}{sep}{variant}'
+        locales = []
+        accept_language = headers.get('Accept-Language', '')
```
Comment (allisonkarlitskaya) [inline src/cockpit/packages.py:187]: I'm reasonably sure you'll want to omit the `''` here...

File: src/cockpit/packages.py [hunk 4]
```diff
@@ -159,12 +183,18 @@ def filename_variants(filename, locale):
...
+        if accept_language:
```
Comment (allisonkarlitskaya) [inline src/cockpit/packages.py:188]: ...and change this to `accept_language is not None`.  No?  Or maybe even one better: 

```python
if isinstance(accept_language, str):
```

File: src/cockpit/packages.py (hunk 2 above)
Comment (allisonkarlitskaya) [reply src/cockpit/packages.py:67]: This function could return empty strings as locales.  It might be worth trying to defend against that.

File: src/cockpit/packages.py [hunk 5]
```diff
@@ -195,6 +225,17 @@ def get_content_security_policy(self, origin):
...
 
+        # HACK: if a translation file is missing, just return empty
+        # content. This saves a whole lot of 404s in the developer
+        # console when trying to fetch po.js for English, for example.
+        if filename is None and path.endswith('po.js'):
+            headers = {
+                "Access-Control-Allow-Origin": channel.origin,
+            }
+            channel.http_ok("application/javascript", headers)
+            channel.send_data(b"")
```
Comment (allisonkarlitskaya) [inline src/cockpit/packages.py:236]: Is that necessary?

File: src/cockpit/packages.py [hunk 6]
```diff
@@ -311,7 +352,10 @@ def serve_checksum(self, channel):
     def serve_file(self, path, channel):
         assert path[0] == '/'
 
-        if self.checksum is not None:
+        # HACK: If the response is language specific, don't cache the file. Caching "po.js" breaks
```
Comment (allisonkarlitskaya) [inline src/cockpit/packages.py:355]: I guess we should also do this for the missing-file (ie: English) case above, no?  Otherwise changing to/from English might be broken.

File: test/pytest/test_packages.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import pytest
+
+from cockpit.packages import parse_accept_language
+
+
+TEST_ACCEPT_LANGUAGES = [
+    ('de-at, zh-CH, en', [('de-at', '1.0'), ('zh-CH', '0.99'), ('en', '0.98')]),
+    ('es-es, nl;q=0.8, fr;q=0.9', [('es-es', '1.0',), ('fr', '0.9'), ('nl', '0.8')]),
+    ('fr-CH, fr;q=0.9, en;q=0.8, de;q=0.7, *;q=0.5', [('fr-CH', '1.0'), ('fr', '0.9'), ('en', '0.8'), ('de', '0.7'), ('*', '0.5')])
+]
```
Comment (allisonkarlitskaya) [inline test/pytest/test_packages.py:27]: This seems less like "useful global constant" and more like data specific to a particular test case.  It belongs inside of that case, I'd say.

File: test/pytest/test_packages.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+
+@pytest.mark.parametrize("test_input,expected", TEST_ACCEPT_LANGUAGES)
```
Comment (allisonkarlitskaya) [inline test/pytest/test_packages.py:30]: ...and I feel like this is probably over-parameterized.  Let's have @martinpitt referee on this one.

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
Thanks for taking care of this.  In general, I think it's time we tried to move towards a slightly more defensive style of programming in the bridge, so see the comments below about checking some error cases.

**Comment by martinpitt** [comment]
> PS: How the hell does this globbing magic even work of loading "/po/*/po.js", does it really concatenate all po files together for all "applications"? I guess I rather not found out

Right, that's exactly the reason why we want to get rid of it and not support globbing in the new bridge :grin:

File: src/cockpit/packages.py [hunk 9]
```diff
@@ -146,7 +170,7 @@ def filename_variants(filename, locale):
                 yield f'{base}.{locale}.{ext}.gz'
 
                 if '_' in locale:
-                    language, _, _ = locale.partition(' ')
+                    language, _, _ = locale.partition('_')
```
Comment (martinpitt) [inline src/cockpit/packages.py:173]: Is that really right? It looks correct for most locales like `de`, but we need to retain the country for at least en, nb, pt, and zh. In general, we should prefer `ll_CC.po` if we have it, and if not fall back to to `ll.po`.

Also, personally I prefer `language = locale.split('_')[0]`, but no strong opinion.

File: test/pytest/test_packages.py (hunk 8 above)
Comment (martinpitt) [reply test/pytest/test_packages.py:30]: Not sure what you mean here, but three scenarios feels quite okay, and they do cover a lot of corner cases.

File: src/cockpit/packages.py (hunk 9 above)
Comment (allisonkarlitskaya) [reply src/cockpit/packages.py:173]: Look up to the line just before the `if`.  We check the full locale name first.  The stripped down version comes next.

Comment (martinpitt) [reply src/cockpit/packages.py:173]: I see, thanks! So @jelly , please feel free to resolve this or replace the partition with a simpler split()[0].

File: src/cockpit/packages.py (hunk 5 above)
Comment (jelly) [reply src/cockpit/packages.py:236]: This needs to be discussed, it seemed irrelevant as a default make generates translations. So maybe this is for cockpit-podman? I rather retain the behaviour and later rip it out.

File: src/cockpit/packages.py (hunk 6 above)
Comment (jelly) [reply src/cockpit/packages.py:355]: Hmm good point! Btw noticed that almost everything is served with no-cache now.

File: test/pytest/test_packages.py (hunk 7 above)
Comment (jelly) [reply test/pytest/test_packages.py:27]: I can move it to parametrize() if you don't want the constant. Basically this is a pytest convenience feature

File: src/cockpit/packages.py (hunk 4 above)
Comment (jelly) [reply src/cockpit/packages.py:188]: Brrr, I rather have None then.

File: src/cockpit/packages.py (hunk 1 above)
Comment (jelly) [reply src/cockpit/packages.py:59]: Right that's the else case.

File: src/cockpit/packages.py (hunk 2 above)
Comment (jelly) [reply src/cockpit/packages.py:67]: * Sure, can drop the tuples
* empty strings as locales, hmm requires a test?

File: src/cockpit/packages.py (hunk 5 above)
Comment (allisonkarlitskaya) [reply src/cockpit/packages.py:238]: I meant only the one line — sending the empty data.  I think that's a no-op and can be excluded — but see above.

File: src/cockpit/packages.py [hunk 10]
```diff
@@ -195,6 +226,18 @@ def get_content_security_policy(self, origin):
     def serve_file(self, path, channel):
         filename = self.negotiate_file(path, channel.headers)
 
+        # HACK: if a translation file is missing, just return empty
+        # content. This saves a whole lot of 404s in the developer
+        # console when trying to fetch po.js for English, for example.
+        if filename is None and path.endswith('po.js'):
+            headers = {
+                'Access-Control-Allow-Origin': channel.origin,
+                'Cache-Control': 'no-cache, no-store',
```
Comment (allisonkarlitskaya) [inline src/cockpit/packages.py:235]: Not a huge fan of the duplication here.

Comment (allisonkarlitskaya) [reply src/cockpit/packages.py:235]: In general I feel like it would be better to handle this case by using the same serving code as you use below, but figuring out a way to serve the empty string instead of the contents of a file.  Something like the moral equivalent of `filename = '/dev/null'`, but without actually reading the file.

File: src/cockpit/packages.py [hunk 11]
```diff
@@ -159,12 +184,18 @@ def filename_variants(filename, locale):
...
-        for variant in self.filename_variants(filename, locale):
-            logger.debug('consider variant %s', variant)
-            if Path(f'{dirname}{sep}{variant}') in self.files:
-                return f'{dirname}{sep}{variant}'
+        locales = []
+        accept_language = headers.get('Accept-Language')
+        if accept_language is not None:
+            locales = parse_accept_language(accept_language)
+
+        for locale in locales:
```
Comment (allisonkarlitskaya) [inline src/cockpit/packages.py:192]: This logic is definitely some kind of wrong.

 - calling the variant function multiple times means that you consider the untranslated filename over and over again
 - doing a separate scan for each page means that we can end up with a mix of languages

The C bridge tries to set the language to the first supported locale it finds, then ignores all of the others.  I wonder if we should do something like that.

If not, and you really want to check multiple languages, then I'd rather modify the filename_variants function to take a list of locales and iterate over them internally before trying the untranslated variants.

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]

File: src/cockpit/packages.py [hunk 12]
```diff
@@ -24,6 +24,7 @@
 import os
 
 from pathlib import Path
+from typing import List, Tuple
```
[the remark this answers is not in the document]
Comment (jelly) [reply src/cockpit/packages.py:27]: I should have a pre-push hook...

File: src/cockpit/packages.py [hunk 13]
```diff
@@ -43,6 +44,30 @@ def directory_items(path):
...
+        weight = float(weight or 1)
+        locale = locale.strip()
+
+        # Skip possible empty locales
+        if not locale:
+            continue
+
+        locales.append((locale, weight))
+
+    return [loc[0] for loc in (sorted(locales, key=lambda k: k[1], reverse=True))]
```
[the remark this answers is not in the document]
Comment (jelly) [reply src/cockpit/packages.py:68]: And no I prefer _ over _weight

File: src/cockpit/packages.py (hunk 10 above)
Comment (jelly) [reply src/cockpit/packages.py]: I did a thing to reduce duplucation.

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
See comments below.

One more extremely annoying point (and I think C bridge also gets this wrong): according to RFC[^1] 

> Language tags and thus language ranges are to be treated as case-insensitive: there exist conventions for the capitalization of some of the subtags, but these MUST NOT be taken to carry meaning.  Matching of language tags to language ranges MUST be done in a case-insensitive manner.

So it seems that if the user gives us `pt-br` or `EN`, we ought to still be able to find the files on our case-sensitive filesystem...  Ugh.

[^1]: https://www.rfc-editor.org/rfc/rfc4647#section-2

File: src/cockpit/packages.py [hunk 14]
```diff
@@ -191,10 +213,20 @@ def filename_variants(filename, locale):
...
-        locale = headers.get('Accept-Language', '').split(',')[0].strip()  # obviously fix this
-
-        for variant in self.filename_variants(filename, locale):
-            logger.debug('consider variant %s', variant)
+        locales = []
+        accept_language = headers.get('Accept-Language')
+        if accept_language is not None:
+            locales = parse_accept_language(accept_language)
+            # Append variants
+            for locale in locales:
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply src/cockpit/packages.py:221]: I was curious, so I wrote that:

```python
seq = [0, 1]  
last = None  
for num in seq:  
    print(num)   
    if last is not None:  
        seq.append(num+last)  
    last = num                
```

and indeed, it loops forever.  Please fix this. :)

File: src/cockpit/packages.py [hunk 15]
```diff
@@ -70,6 +71,30 @@ def get_libexecdir() -> str:
...
+
+    accept_languages = accept_language.split(',')
+    locales = []
+    for language in accept_languages:
+        language = language.strip()
+        locale, _, weight = language.partition(';q=')
+        weight = float(weight or 1)
+
+        # Skip possible empty locales
+        if not locale:
```
Comment (allisonkarlitskaya) [inline src/cockpit/packages.py:90]: :+1:  Thanks for this addition.

File: src/cockpit/packages.py [hunk 16]
```diff
@@ -227,22 +259,31 @@ def get_content_security_policy(self, origin):
     def serve_file(self, path, channel):
         filename = self.negotiate_file(path, channel.headers)
 
-        if filename is None:
+        if filename:
```
Comment (allisonkarlitskaya) [inline src/cockpit/packages.py:262]: `... is not None` lol

I don't think it matters in this case, because there's no way we found an empty filename as a variant, but seriously :sweat_smile:

File: src/cockpit/packages.py [hunk 17]
```diff
@@ -227,22 +259,31 @@ def get_content_security_policy(self, origin):
...
+            with (self.path / filename).open('rb') as file:
+                data = file.read()
+        # HACK: if a translation file is missing, just return empty
+        # content. This saves a whole lot of 404s in the developer
+        # console when trying to fetch po.js for English, for example.
+        elif path.endswith('po.js'):
+            data = b''
+            filename = 'po.js'
+        else:
+            logging.error("filename=%s, path=%s", filename, path)
```
Comment (allisonkarlitskaya) [inline src/cockpit/packages.py:272]: This arrangement is *so* much nicer.  Thanks!

File: src/cockpit/packages.py [hunk 18]
```diff
@@ -227,22 +259,31 @@ def get_content_security_policy(self, origin):
     def serve_file(self, path, channel):
         filename = self.negotiate_file(path, channel.headers)
 
-        if filename is None:
+        if filename:
+            with (self.path / filename).open('rb') as file:
+                data = file.read()
+        # HACK: if a translation file is missing, just return empty
```
Comment (allisonkarlitskaya) [inline src/cockpit/packages.py:265]: I think it reads better if this is inside the elif block, but it's a matter of taste.

**Review by allisonkarlitskaya — APPROVED** [review]
This looks good now, thanks :)
